modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxUnlockRequestSender.java

```diff
@@ -0,0 +1,171 @@
+     */
+    public TxUnlockRequestSender(
+            ClusterService clusterService,
+            PlacementDriverHelper placementDriverHelper,
+            HybridClock clock,
+            TxCleanupProcessor txCleanupProcessor) {
+        this.clusterService = clusterService;
+        this.placementDriverHelper = placementDriverHelper;
+        this.hybridClock = clock;
+        this.txCleanupProcessor = txCleanupProcessor;
+    }
+
+    /**
+     * Sends unlock request to the primary nodes of each one of {@code partitions}.
+     *
+     * @param partitions Enlisted partition groups.
+     * @param commit {@code true} if a commit requested.
+     * @param commitTimestamp Commit timestamp ({@code null} if it's an abort).
+     * @param txId Transaction id.
+     * @return Completable future of Void.
+     */
+    public CompletableFuture<Void> unlock(
+            Collection<TablePartitionId> partitions,
+            boolean commit,
+            @Nullable HybridTimestamp commitTimestamp,
+            UUID txId
+    ) {
+        return placementDriverHelper.findPrimaryReplicas(partitions, hybridClock.now())
+                .thenCompose(partitionData -> {
+                    durableCleanupPartitionsWithNoPrimary(commit, commitTimestamp, txId, partitionData.partitionsWithoutPrimary);
+
+                    return unlockPartitions(partitionData.partitionsByNode, commit, commitTimestamp, txId);
+                });
+    }
+
+    private void durableCleanupPartitionsWithNoPrimary(
+            boolean commit,
+            @Nullable HybridTimestamp commitTimestamp,
+            UUID txId,
+            Set<TablePartitionId> noPrimaryFound
+    ) {
+        if (!noPrimaryFound.isEmpty()) {
+            for (TablePartitionId partition : noPrimaryFound) {
+                // Okay, no primary found for that partition.
+                // Switch to the durable cleanup that will wait for the primary to appear.
+                // Also no need to wait on this future - if the primary has expired, the locks are already released.
+                txCleanupProcessor.cleanupWithRetry(commit, commitTimestamp, txId, partition);
+            }
+        }
+    }
+
+    private CompletableFuture<Void> unlockPartitions(
+            Map<String, Set<TablePartitionId>> partitionsByNode,
+            boolean commit,
+            @Nullable HybridTimestamp commitTimestamp,
+            UUID txId
+    ) {
+        List<CompletableFuture<Void>> unlockFutures = new ArrayList<>();
+
+        for (Entry<String, Set<TablePartitionId>> entry : partitionsByNode.entrySet()) {
+            String node = entry.getKey();
+            Set<TablePartitionId> nodePartitions = entry.getValue();
+
+            unlockFutures.add(sendUnlockMessageWithRetries(commit, commitTimestamp, txId, node, nodePartitions));
+        }
+
+        return allOf(unlockFutures.toArray(new CompletableFuture<?>[0]));
+    }
+
+    private CompletableFuture<Void> sendUnlockMessageWithRetries(
```

Comment:
It's also a bit confusing. From the unlock() method point of view cleanup is durable and unlock is not, however unlock also occurs to be durable but the fact that it is - is covered with different terms - withRetries and within different layer. Let's be consistent here.